dataplane/README.md

```diff
@@ -0,0 +1,176 @@
+# The Lemming Dataplane
+
+The dataplane reads and writes packets from ports and the forwards packets between them.
+
+## Overview
+
+The dataplane consists of a several components:
+
+1. forwarding: packet processing actions, tables, and ports.
+1. saiserver: implements a "proto-ized" version of the [Switch Abstraction Interface API](https://github.com/opencomputeproject/SAI) and defines the forwarding pipeline
+1. apigen: code to generate the proto API for the saiserver and C++ code that calls the API
+1. dplanerc: gNMI reconciler that configures the dataplane based on OpenConfig schema
+1. standalone: shared library that a implements the C SAI API and container running only the dataplane
+
+There are two ways to run the dataplane. The first as part of Lemming, using the gNMI reconciliation. The second is a SAI implementation (e.g. in SONIC).
+
+## Components
+
+### forwarding
+
+The forwarding directory contains packet processing actions, tables, and ports. These are the building blocks of the dataplane: they can be combined to create the
+packet processing pipeline. The forwarding behavior is configured using the [gRPC API](../proto/forwarding/forwarding_service.proto).
+
+Generally, external users should not program the dataplane using this API directly. It is too flexible and low-table. Instead, the saiserver API should be used instead.
+
+Each port has a stack of input and output actions. The forwarding server processes an incoming packet using the input actions of the output.
+If after processing all the actions, there is an output port, then the server processes the output actions of the output port.
+
+The following sections describe the key parts of the forwarding server. The relevant packages and protobuf files contain more complete documentation.
+
+#### Actions
+
+Actions can modify the contents of the packet, output the packet etc...
+
+Some the important actions include:
+
+* Update: Set packet headers values
+* Transmit: Set the output port
+* Lookup: Lookup in a table
```

Comment:
Also mention that a table can act on the packet with other actions (see tables heading)